core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -556,6 +549,42 @@ public Object readGenericValue() throws IOException {
         return null;
     }
 
+    /**
+     * Reads a {@link NamedWriteable} from the current stream, by first reading its name and then looking for
+     * the corresponding entry in the registry by name, so that the proper object can be read and returned.
+     * Default implementation throws {@link UnsupportedOperationException} as StreamInput doesn't hold a registry.
+     * Use {@link FilterInputStream} instead which wraps a stream and supports a {@link NamedWriteableRegistry} too.
+     */
+    <C> C readNamedWriteable(@SuppressWarnings("unused") Class<C> categoryClass) throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+
+    /**
+     * Reads an optional {@link NamedWriteable} from the current stream, by first reading its name and then looking for
+     * the corresponding entry in the registry by name, so that the proper object can be read and returned.
+     */
+    final <C> C readOptionalNamedWriteable(Class<C> categoryClass) throws IOException {
+        if (readBoolean()) {
+            return readNamedWriteable(categoryClass);
+        }
+        return null;
+    }
+
+    /**
+     * Reads a list of {@link NamedWriteable} from the current stream, by first reading its size and then
+     * reading the individual objects using {@link #readNamedWriteable(Class)}.
+     */
+    final <C> List<C> readNamedWriteableList(Class<C> categoryClass) throws IOException {
+        List<C> list = new ArrayList<>();
+        int size = readInt();
+        for (int i = 0; i < size; i++) {
+            C obj = readNamedWriteable(categoryClass);
+            list.add(obj);
+        }
+        return list;
+    }
```

Comment:
yep that was the idea, it is confusing cause they are only used for testing.... prod code that uses them will come later.